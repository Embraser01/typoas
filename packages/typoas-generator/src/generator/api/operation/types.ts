import {
  ParameterObject,
  ReferenceObject,
  SecurityRequirementObject,
} from 'openapi3-ts/oas31';

export type GlobalParameters = {
  baseParameters: (ParameterObject | ReferenceObject)[];
  defaultSecurityRequirements?: SecurityRequirementObject[];
};
