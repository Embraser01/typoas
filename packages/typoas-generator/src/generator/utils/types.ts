import { ReferenceObject, SchemaObject } from 'openapi3-ts';
import { factory, TypeNode, SyntaxKind } from 'typescript';
import { Context } from '../../context';
import {
  hasUnsupportedIdentifierChar,
  sanitizeTypeIdentifier,
} from './operation-name';

const ENUM_SUPPORTED_TYPE = ['integer', 'number', 'string', 'null', 'boolean'];

export function createTypeFromSchema(
  schemaOrRef: SchemaObject | ReferenceObject | undefined,
  ctx: Context,
): TypeNode {
  let node: TypeNode = factory.createKeywordTypeNode(SyntaxKind.AnyKeyword);

  if (!schemaOrRef) {
    return node;
  }
  if (schemaOrRef.$ref) {
    const ref = ctx.resolveReference('schemas', schemaOrRef.$ref);
    if (!ref) {
      throw new Error(`$ref '${schemaOrRef.$ref}' wasn't found`);
    }
    node = factory.createTypeReferenceNode(
      factory.createIdentifier(sanitizeTypeIdentifier(ref.name)),
    );
  }

  const schema: SchemaObject = schemaOrRef;
  if (schema.enum && schema.type && ENUM_SUPPORTED_TYPE.includes(schema.type)) {
    node = factory.createUnionTypeNode(
      schema.enum
        .map((e) => {
          if (e === null) return factory.createNull();
          if (typeof e === 'boolean') {
            return e ? factory.createTrue() : factory.createFalse();
          }
          if (typeof e === 'number') {
            return factory.createNumericLiteral(e);
          }
          return factory.createStringLiteral(e, true);
        })
        .map((e) => factory.createLiteralTypeNode(e)),
    );
  } else if (schema.oneOf && schema.oneOf.length) {
    node = factory.createUnionTypeNode(
      schema.oneOf.map((s) => createTypeFromSchema(s, ctx)),
    );
  } else if (schema.anyOf && schema.anyOf.length) {
    node = factory.createUnionTypeNode(
      schema.anyOf.map((s) => createTypeFromSchema(s, ctx)),
    );
  } else {
    switch (schema.type) {
      case 'integer':
      case 'number':
        node = factory.createKeywordTypeNode(SyntaxKind.NumberKeyword);
        break;
      case 'string':
        switch (schema.format) {
          case 'date':
          case 'date-time':
            node = factory.createTypeReferenceNode(
              factory.createIdentifier('Date'),
            );
            break;
          default:
            node = factory.createKeywordTypeNode(SyntaxKind.StringKeyword);
            break;
        }
        break;
      case 'boolean':
        node = factory.createKeywordTypeNode(SyntaxKind.BooleanKeyword);
        break;
      case 'object': {
        const hasProperties =
          schema.properties && Object.keys(schema.properties).length;

        if (hasProperties) {
          node = factory.createTypeLiteralNode(
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            Object.entries(schema.properties!).map(([key, val]) =>
              factory.createPropertySignature(
                undefined,
                hasUnsupportedIdentifierChar(key)
                  ? factory.createStringLiteral(key, true)
                  : factory.createIdentifier(key),
                schema.required?.includes(key)
                  ? undefined
                  : factory.createToken(SyntaxKind.QuestionToken),
                createTypeFromSchema(val, ctx),
              ),
            ),
          );
        }
        if (schema.additionalProperties) {
          const additionalPropsNode = factory.createTypeLiteralNode([
            factory.createIndexSignature(
              undefined,
              undefined,
              [
                factory.createParameterDeclaration(
                  undefined,
                  undefined,
                  undefined,
                  factory.createIdentifier('key'),
                  undefined,
                  factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
                  undefined,
                ),
              ],
              schema.additionalProperties === true
                ? factory.createKeywordTypeNode(SyntaxKind.AnyKeyword)
                : createTypeFromSchema(schema.additionalProperties, ctx),
            ),
          ]);

          node = hasProperties
            ? factory.createIntersectionTypeNode([node, additionalPropsNode])
            : additionalPropsNode;
        }
        break;
      }
      case 'null':
        node = factory.createLiteralTypeNode(factory.createNull());
        break;
      case 'array':
        node = factory.createArrayTypeNode(
          createTypeFromSchema(schema.items, ctx),
        );
        break;
    }
  }

  if (schema.nullable) {
    node = factory.createUnionTypeNode([
      node,
      factory.createLiteralTypeNode(factory.createNull()),
    ]);
  }
  return node;
}
