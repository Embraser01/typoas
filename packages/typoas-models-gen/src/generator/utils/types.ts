import { ReferenceObject, SchemaObject } from 'openapi3-ts';
import { factory, TypeNode, SyntaxKind } from 'typescript';
import { Context } from '../../context';

export function createTypeFromSchema(
  schemaOrRef: SchemaObject | ReferenceObject | undefined,
  ctx: Context,
): TypeNode {
  if (!schemaOrRef) {
    return factory.createKeywordTypeNode(SyntaxKind.AnyKeyword);
  }
  if (schemaOrRef.$ref) {
    const ref = ctx.resolveReference('schemas', schemaOrRef.$ref);
    if (!ref) {
      throw new Error(`$ref '${schemaOrRef.$ref}' wasn't found`);
    }
    return factory.createTypeReferenceNode(factory.createIdentifier(ref.name));
  }

  const schema: SchemaObject = schemaOrRef;

  switch (schema.type) {
    case 'integer':
    case 'number':
      return factory.createKeywordTypeNode(SyntaxKind.NumberKeyword);
    case 'string':
      switch (schema.format) {
        case 'date':
        case 'date-time':
          return factory.createTypeReferenceNode(
            factory.createIdentifier('Date'),
          );
        default:
          return factory.createKeywordTypeNode(SyntaxKind.StringKeyword);
      }
    case 'boolean':
      return factory.createKeywordTypeNode(SyntaxKind.BooleanKeyword);
    case 'object': {
      let node: TypeNode = factory.createKeywordTypeNode(SyntaxKind.AnyKeyword);
      if (schema.properties && Object.keys(schema.properties).length) {
        node = factory.createTypeLiteralNode(
          Object.entries(schema.properties).map(([key, val]) =>
            factory.createPropertySignature(
              undefined,
              factory.createIdentifier(key),
              schema.required?.includes(key)
                ? undefined
                : factory.createToken(SyntaxKind.QuestionToken),
              createTypeFromSchema(val, ctx),
            ),
          ),
        );
      }
      if (schema.additionalProperties) {
        const additionalPropsNode = factory.createTypeLiteralNode([
          factory.createIndexSignature(
            undefined,
            undefined,
            [
              factory.createParameterDeclaration(
                undefined,
                undefined,
                undefined,
                factory.createIdentifier('key'),
                undefined,
                factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
                undefined,
              ),
            ],
            schema.additionalProperties === true
              ? factory.createKeywordTypeNode(SyntaxKind.AnyKeyword)
              : createTypeFromSchema(schema.additionalProperties, ctx),
          ),
        ]);

        node = !node
          ? additionalPropsNode
          : factory.createIntersectionTypeNode([node, additionalPropsNode]);
      }
      return node;
    }
    case 'null':
      return factory.createLiteralTypeNode(factory.createNull());
    case 'array':
      return factory.createArrayTypeNode(
        createTypeFromSchema(schema.items, ctx),
      );
  }

  if (schema.oneOf && schema.oneOf.length) {
    return factory.createUnionTypeNode(
      schema.oneOf.map((s) => createTypeFromSchema(s, ctx)),
    );
  }

  if (schema.anyOf && schema.anyOf.length) {
    return factory.createUnionTypeNode(
      schema.anyOf.map((s) => createTypeFromSchema(s, ctx)),
    );
  }

  return factory.createKeywordTypeNode(SyntaxKind.AnyKeyword);
}
