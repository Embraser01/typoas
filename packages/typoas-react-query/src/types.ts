import {
  EnhancedHTTPStatus,
  SpecificStatusResponse,
  StatusResponse,
} from '@typoas/runtime';

export type TypoasReturnType<
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Func extends (...args: any[]) => Promise<any>,
  S extends EnhancedHTTPStatus,
> = SpecificStatusResponse<Awaited<ReturnType<Func>>, S>;

export type TypoasFuncStatusType<
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Func extends (...args: any[]) => Promise<any>,
> = Awaited<ReturnType<Func>> extends StatusResponse<infer S> ? S : never;
