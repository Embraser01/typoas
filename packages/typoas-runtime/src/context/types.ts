import type { Fetcher, HttpMethod, SerializerOptions } from '../fetcher';
import type { TransformField } from '../transformers';
import type { SecurityAuthentication } from '../auth';
import type { TransformResolver } from '../resolver';
import type { Transform } from '../transformers';
import type { BaseServerConfiguration } from '../configuration';

export type CreateRequestParams = {
  /**
   * Method used by the request.
   */
  method: HttpMethod;
  /**
   * Raw path of the request. Params needed to be added to the path
   * will be added by the context.
   */
  path: string;
  /**
   * Params that will be used to be added to path or query params.
   */
  params: Record<string, unknown>;
  /**
   * Headers to be added to the request.
   */
  headers?: Record<string, unknown>;
  /**
   * Body of the request.
   */
  body?: unknown;
  /**
   * List of query params to be taken from params.
   */
  queryParams?: string[];
  /**
   * Name of the auth modes to be used.
   */
  auth?: string[];
};

export type ContextParams = {
  resolver: TransformResolver;
  serverConfiguration: BaseServerConfiguration;
  fetcher?: Fetcher;
  transformers?: Record<string, Transform<unknown, unknown>>;
  serializerOptions?: SerializerOptions;
  authMethods?: Record<string, SecurityAuthentication>;
};

export type ResponseHandler = {
  success?: boolean;
  transforms?: Record<string, TransformField>;
};
