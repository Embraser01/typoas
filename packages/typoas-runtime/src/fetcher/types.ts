/**
 * Represents an HTTP method.
 */
export enum HttpMethod {
  GET = 'GET',
  HEAD = 'HEAD',
  POST = 'POST',
  PUT = 'PUT',
  DELETE = 'DELETE',
  CONNECT = 'CONNECT',
  OPTIONS = 'OPTIONS',
  TRACE = 'TRACE',
  PATCH = 'PATCH',
}

/**
 * Represents an HTTP file which will be transferred from or to a server.
 */
export type HttpFile = Blob & { readonly name: string };

/**
 * Represents the body of an outgoing HTTP request.
 */
export type RequestBody =
  | undefined
  | string
  | FormData
  | Blob
  | URLSearchParams;

export type QueryStyles =
  | 'form'
  | 'spaceDelimited'
  | 'pipeDelimited'
  | 'deepObject';
/**
 * Options are following https://swagger.io/docs/specification/serialization/.
 */
export type SerializerOptions = {
  explode?: boolean;
  queryStyle?: QueryStyles;
};

export interface ResponseBody {
  text(): Promise<string>;
  binary(): Promise<Blob>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  json(): Promise<any>;
}

/**
 * Represents an HTTP status code including the default
 * status codes and the 1XX, 4XX, 5XX ranges.
 */
export type EnhancedHTTPStatus =
  | 'default'
  | '1XX'
  | '2XX'
  | '3XX'
  | '4XX'
  | '5XX'
  | 100
  | 101
  | 102
  | 103
  | 200
  | 201
  | 202
  | 203
  | 204
  | 205
  | 206
  | 207
  | 208
  | 226
  | 300
  | 301
  | 302
  | 303
  | 304
  | 305
  | 306
  | 307
  | 308
  | 400
  | 401
  | 402
  | 403
  | 404
  | 405
  | 406
  | 407
  | 408
  | 409
  | 410
  | 411
  | 412
  | 413
  | 414
  | 415
  | 416
  | 417
  | 418
  | 421
  | 422
  | 423
  | 424
  | 425
  | 426
  | 427
  | 428
  | 429
  | 430
  | 431
  | 451
  | 500
  | 501
  | 502
  | 503
  | 504
  | 505
  | 506
  | 507
  | 508
  | 509
  | 510
  | 511;

/**
 * Extracts only the successful status codes of EnhancedHTTPStatus.
 */
export type SuccessfulStatus = Extract<
  EnhancedHTTPStatus,
  | 'default'
  | '2XX'
  | 200
  | 201
  | 202
  | 203
  | 204
  | 205
  | 206
  | 207
  | 208
  | 226
  | '3XX'
  | 300
  | 301
  | 302
  | 303
  | 304
  | 305
  | 306
  | 307
  | 308
>;

/**
 * Represents a response from an api call.
 */
export type StatusResponse<
  S extends EnhancedHTTPStatus = 'default',
  R = unknown,
> = {
  data: R;
  // Here we don't use the Header type as not all fetchers will have the same headers type.
  headers: Record<string, string>;
  ok: S extends SuccessfulStatus ? true : S extends 'default' ? boolean : false;
  // When using ranges or default, we can't now the specific status code.
  status: S extends string ? number : S;
};
