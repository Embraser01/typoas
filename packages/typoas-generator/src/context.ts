/* eslint-disable @typescript-eslint/no-explicit-any */
import { TypeNode } from 'typescript';
import {
  CallbackObject,
  ComponentsObject,
  ExampleObject,
  HeaderObject,
  LinkObject,
  ParameterObject,
  ReferenceObject,
  RequestBodyObject,
  ResponseObject,
  SchemaObject,
  SecuritySchemeObject,
} from 'openapi3-ts';

type ComponentContext<T> = {
  tsNode?: TypeNode;
  imports: string[];
  name: string;
  spec: T;
};

export type ComponentType =
  | 'schemas'
  | 'responses'
  | 'parameters'
  | 'examples'
  | 'requestBodies'
  | 'headers'
  | 'securitySchemes'
  | 'links'
  | 'callbacks';

export type ComponentContextFromName<T extends ComponentType> =
  T extends 'schemas'
    ? ComponentContext<SchemaObject>
    : T extends 'responses'
    ? ComponentContext<ResponseObject>
    : T extends 'parameters'
    ? ComponentContext<ParameterObject>
    : T extends 'examples'
    ? ComponentContext<ExampleObject>
    : T extends 'requestBodies'
    ? ComponentContext<RequestBodyObject>
    : T extends 'headers'
    ? ComponentContext<HeaderObject>
    : T extends 'securitySchemes'
    ? ComponentContext<SecuritySchemeObject>
    : T extends 'links'
    ? ComponentContext<LinkObject>
    : ComponentContext<CallbackObject>;

export type ComponentRegistry<T extends ComponentType> = Map<
  string,
  ComponentContextFromName<T>
>;

export type ContextOptions = {
  generateEnums?: boolean;
  jsDoc?: boolean;
  onlyTypes?: boolean;
  anyInsteadOfUnknown?: boolean;
};

export class Context {
  schemas: ComponentRegistry<'schemas'> = new Map();
  responses: ComponentRegistry<'responses'> = new Map();
  parameters: ComponentRegistry<'parameters'> = new Map();
  examples: ComponentRegistry<'examples'> = new Map();
  requestBodies: ComponentRegistry<'requestBodies'> = new Map();
  headers: ComponentRegistry<'headers'> = new Map();
  securitySchemes: ComponentRegistry<'securitySchemes'> = new Map();
  links: ComponentRegistry<'links'> = new Map();
  callbacks: ComponentRegistry<'callbacks'> = new Map();

  constructor(private opts: ContextOptions = {}) {}

  resolveReference<T extends ComponentType>(
    type: T,
    ref: string,
  ): ComponentContextFromName<T> | null {
    const [, schemaName] =
      new RegExp(`^#/components/${type}/([^/]+)`).exec(ref) || [];

    let map;
    switch (type) {
      case 'schemas':
        map = this.schemas;
        break;
      case 'responses':
        map = this.responses;
        break;
      case 'parameters':
        map = this.parameters;
        break;
      case 'examples':
        map = this.examples;
        break;
      case 'requestBodies':
        map = this.requestBodies;
        break;
      case 'headers':
        map = this.headers;
        break;
      case 'securitySchemes':
        map = this.securitySchemes;
        break;
      case 'links':
        map = this.links;
        break;
      case 'callbacks':
        map = this.callbacks;
        break;
    }
    const schema = (map as ComponentRegistry<T>).get(schemaName);
    if (!schema) return null;
    return schema;
  }

  private addToMap(
    name: string,
    spec: ReferenceObject | any,
    map: ComponentRegistry<any>,
  ): void {
    if (map.has(name)) return;

    if (spec.$ref) {
      // TODO Warning
    }
    map.set(name, { imports: [], name, spec });
  }

  initComponents(components: ComponentsObject): void {
    Object.entries(components).forEach(([type, value]) =>
      Object.entries(value).forEach(([name, spec]) =>
        // @ts-expect-error This[type] is not handled by TS
        this.addToMap(name, spec, this[type]),
      ),
    );
  }

  generateEnums(): boolean {
    return this.opts.generateEnums === true;
  }

  generateAnyInsteadOfUnknown(): boolean {
    return this.opts.anyInsteadOfUnknown === true;
  }

  hasJSDoc(): boolean {
    return this.opts.jsDoc !== false;
  }

  isOnlyTypes(): boolean {
    return !!this.opts.onlyTypes;
  }
}
