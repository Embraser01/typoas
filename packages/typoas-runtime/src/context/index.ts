import type {
  ContextParams,
  CreateRequestParams,
  ResponseHandler,
} from './types';
import type { SecurityAuthentication } from '../auth';
import type { BaseServerConfiguration } from '../configuration';
import { ApiException } from '../exception';
import { applyTransform, Transform, TransformResolver } from '../transformers';
import { DateTransformer } from '../transformers';
import {
  Fetcher,
  IsomorphicFetchHttpLibrary,
  RequestContext,
  ResponseContext,
  SerializerOptions,
} from '../fetcher';
import {
  applyTemplating,
  isCodeInRange,
  serializeHeader,
  serializeParameter,
} from '../utils';

export class Context<AuthModes extends Record<string, SecurityAuthentication>> {
  fetcher: Fetcher;
  serializerOptions: SerializerOptions;
  authMethods: Partial<AuthModes>;
  resolver: TransformResolver;
  transformers: Record<string, Transform<unknown, unknown>>;
  serverConfiguration: BaseServerConfiguration;

  constructor(params: ContextParams<AuthModes>) {
    this.resolver = params.resolver;
    this.serverConfiguration = params.serverConfiguration;
    this.fetcher = params.fetcher || new IsomorphicFetchHttpLibrary();
    this.serializerOptions = params.serializerOptions || {};
    this.authMethods = params.authMethods;
    this.transformers = params.transformers || { date: DateTransformer };
  }

  /**
   * Creates a request context ready to be sent.
   */
  async createRequest(options: CreateRequestParams): Promise<RequestContext> {
    const path = applyTemplating(options.path, options.params);

    const requestContext = new RequestContext(
      this.serverConfiguration.getURL(path),
      options.method,
      this.serializerOptions,
    );

    if (options.headers) {
      for (const [key, value] of Object.entries(options.headers)) {
        requestContext.setHeaderParam(key, serializeHeader(value));
      }
    }
    if (options.queryParams?.length) {
      for (const queryParam of options.queryParams) {
        if (options.params[queryParam] !== undefined) {
          requestContext.setQueryParam(
            queryParam,
            serializeParameter(options.params[queryParam]),
          );
        }
      }
    }
    if (options.body !== undefined) {
      requestContext.setBody(JSON.stringify(options.body));
    }
    if (options.auth?.length) {
      for (const mode of options.auth) {
        await this.authMethods[mode]?.applySecurityAuthentication(
          requestContext,
        );
      }
    }
    return requestContext;
  }

  /**
   * Sends a request.
   */
  sendRequest(request: RequestContext): Promise<ResponseContext> {
    return this.fetcher.send(request);
  }

  /**
   * Handle response by applying transforms.
   */
  async handleResponse<T = unknown>(
    res: ResponseContext,
    handlers: Record<string, ResponseHandler>,
  ): Promise<T> {
    let statusCode = Object.keys(handlers)
      .filter((code) => code !== 'default')
      .find((code) => isCodeInRange(code, res.httpStatusCode));

    if (!statusCode && handlers.default) {
      statusCode = 'default';
    }
    if (!statusCode) {
      throw new ApiException(
        res.httpStatusCode,
        `No handler for status code ${res.httpStatusCode}`,
      );
    }
    const handler = handlers[statusCode];

    const body = {}; // TODO Parse body from response

    if (handler.transforms) {
      for (const [key, transformer] of Object.entries(this.transformers)) {
        const transforms = handler.transforms[key];
        if (transforms) {
          for (const transform of transforms) {
            applyTransform(
              this.resolver,
              { body },
              'body',
              key,
              transformer,
              transform,
              0,
            );
          }
        }
      }
    }

    // Do not throw on valid http status code.
    if (res.httpStatusCode >= 200 && res.httpStatusCode < 400) {
      return body as T;
    }
    throw new ApiException(res.httpStatusCode, body);
  }
}
