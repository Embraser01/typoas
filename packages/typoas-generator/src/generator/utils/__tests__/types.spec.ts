import { describe, expect, it } from '@jest/globals';
import { ReferenceObject, SchemaObject } from 'openapi3-ts';
import {
  createPrinter,
  EmitHint,
  factory,
  NewLineKind,
  SyntaxKind,
  NodeFlags,
  TypeNode,
} from 'typescript';
import { createTypeFromSchema } from '../types';
import { Context } from '../../../context';

const getStringFromNode = (node: TypeNode): string => {
  const printer = createPrinter({ newLine: NewLineKind.LineFeed });
  const resultFile = factory.createSourceFile(
    [],
    factory.createToken(SyntaxKind.EndOfFileToken),
    NodeFlags.Const,
  );

  return printer.printNode(EmitHint.Unspecified, node, resultFile);
};

describe('create type from schema', () => {
  it('should handle integer schema', () => {
    const schema: SchemaObject = { type: 'integer' };

    const node = createTypeFromSchema(schema, new Context());

    expect(getStringFromNode(node)).toMatchSnapshot();
  });

  it('should handle primitive number schema', () => {
    const schema: SchemaObject = { type: 'number' };

    const node = createTypeFromSchema(schema, new Context());

    expect(getStringFromNode(node)).toMatchSnapshot();
  });

  it('should handle primitive string schema', () => {
    const schema: SchemaObject = { type: 'string' };

    const node = createTypeFromSchema(schema, new Context());

    expect(getStringFromNode(node)).toMatchSnapshot();
  });

  it('should handle primitive boolean schema', () => {
    const schema: SchemaObject = { type: 'boolean' };

    const node = createTypeFromSchema(schema, new Context());

    expect(getStringFromNode(node)).toMatchSnapshot();
  });

  it('should handle primitive boolean enums', () => {
    const schema: SchemaObject = { type: 'boolean', enum: [true] };

    const node = createTypeFromSchema(schema, new Context());

    expect(getStringFromNode(node)).toMatchSnapshot();
  });

  it('should handle primitive string enums', () => {
    const schema: SchemaObject = { type: 'string', enum: ['a', 'b'] };

    const node = createTypeFromSchema(schema, new Context());

    expect(getStringFromNode(node)).toMatchSnapshot();
  });

  it('should handle primitive number enums', () => {
    const schema: SchemaObject = { type: 'number', enum: [1, 6] };

    const node = createTypeFromSchema(schema, new Context());

    expect(getStringFromNode(node)).toMatchSnapshot();
  });

  it('should handle dates', () => {
    const schema: SchemaObject = { type: 'string', format: 'date-time' };

    const node = createTypeFromSchema(schema, new Context());

    expect(getStringFromNode(node)).toMatchSnapshot();
  });

  it('should handle array of strings', () => {
    const schema: SchemaObject = { type: 'array', items: { type: 'string' } };

    const node = createTypeFromSchema(schema, new Context());

    expect(getStringFromNode(node)).toMatchSnapshot();
  });

  it('should handle simple object schema', () => {
    const schema: SchemaObject = {
      type: 'object',
      required: ['a'],
      properties: {
        a: { type: 'string' },
        b: { type: 'number' },
      },
    };

    const node = createTypeFromSchema(schema, new Context());
    expect(getStringFromNode(node)).toMatchSnapshot();
  });

  it('should handle object schema without properties', () => {
    const schema: SchemaObject = {
      type: 'object',
      additionalProperties: true,
    };

    const node = createTypeFromSchema(schema, new Context());
    expect(getStringFromNode(node)).toMatchSnapshot();
  });

  describe('with complex object', () => {
    it('should create nested objects', () => {
      const schema: SchemaObject = {
        type: 'object',
        required: ['a'],
        properties: {
          a: { type: 'string' },
          b: { type: 'number' },
          c: {
            type: 'object',
            properties: {
              d: { type: 'string' },
              e: {
                type: 'object',
                properties: {
                  f: { type: 'number' },
                },
              },
            },
          },
        },
      };

      const node = createTypeFromSchema(schema, new Context());
      expect(getStringFromNode(node)).toMatchSnapshot();
    });

    it('should allow additional properties if set to true', () => {
      const schema: SchemaObject = {
        type: 'object',
        required: ['a'],
        additionalProperties: true,
        properties: {
          a: { type: 'string' },
          b: { type: 'number' },
        },
      };

      const node = createTypeFromSchema(schema, new Context());
      expect(getStringFromNode(node)).toMatchSnapshot();
    });

    it('should handle additionalProperties specific type', () => {
      const schema: SchemaObject = {
        type: 'object',
        additionalProperties: {
          type: 'object',
          properties: { a: { type: 'string' } },
        },
      };

      const node = createTypeFromSchema(schema, new Context());
      expect(getStringFromNode(node)).toMatchSnapshot();
    });

    it('should handle anyOf', () => {
      const schema: SchemaObject = {
        anyOf: [
          { type: 'string' },
          { type: 'array', items: { type: 'string' } },
        ],
      };

      const node = createTypeFromSchema(schema, new Context());
      expect(getStringFromNode(node)).toMatchSnapshot();
    });

    it('should handle oneOf', () => {
      const schema: SchemaObject = {
        anyOf: [
          { type: 'string' },
          { type: 'array', items: { type: 'string' } },
        ],
      };

      const node = createTypeFromSchema(schema, new Context());
      expect(getStringFromNode(node)).toMatchSnapshot();
    });
  });

  describe('with refs', () => {
    it('should work referencing ref', () => {
      const ctx = new Context();
      ctx.initComponents({ schemas: { BaseItem: { type: 'string' } } });

      const schema: ReferenceObject = { $ref: '#/components/schemas/BaseItem' };

      const node = createTypeFromSchema(schema, ctx);
      expect(getStringFromNode(node)).toMatchSnapshot();
    });

    it('should work in nested properties', () => {
      const ctx = new Context();
      ctx.initComponents({ schemas: { BaseItem: { type: 'string' } } });

      const schema: SchemaObject = {
        type: 'object',
        properties: {
          a: { $ref: '#/components/schemas/BaseItem' },
        },
      };

      const node = createTypeFromSchema(schema, ctx);
      expect(getStringFromNode(node)).toMatchSnapshot();
    });
  });
});
