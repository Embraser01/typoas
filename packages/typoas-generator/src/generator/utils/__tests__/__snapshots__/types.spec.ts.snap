// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create type from schema should handle array of strings 1`] = `"string[]"`;

exports[`create type from schema should handle const string 1`] = `"'a'"`;

exports[`create type from schema should handle dates 1`] = `"Date"`;

exports[`create type from schema should handle integer schema 1`] = `"number"`;

exports[`create type from schema should handle object schema without properties 1`] = `
"{
    [key: string]: any;
}"
`;

exports[`create type from schema should handle primitive boolean enums 1`] = `"true"`;

exports[`create type from schema should handle primitive boolean schema 1`] = `"boolean"`;

exports[`create type from schema should handle primitive number enums 1`] = `"1 | 6"`;

exports[`create type from schema should handle primitive number schema 1`] = `"number"`;

exports[`create type from schema should handle primitive string enums 1`] = `"'a' | 'b'"`;

exports[`create type from schema should handle primitive string schema 1`] = `"string"`;

exports[`create type from schema should handle simple object schema 1`] = `
"{
    a: string;
    b?: number;
}"
`;

exports[`create type from schema should handle type as an array 1`] = `"string | number"`;

exports[`create type from schema with complex object should allow additional properties if set to true 1`] = `
"{
    a: string;
    b?: number;
} & {
    [key: string]: any;
}"
`;

exports[`create type from schema with complex object should create nested objects 1`] = `
"{
    a: string;
    b?: number;
    c?: {
        d?: string;
        e?: {
            f?: number;
        };
    };
}"
`;

exports[`create type from schema with complex object should handle additionalProperties specific type 1`] = `
"{
    [key: string]: {
        a?: string;
    };
}"
`;

exports[`create type from schema with complex object should handle allOf 1`] = `
"{
    a?: string;
} & {
    b?: number;
}"
`;

exports[`create type from schema with complex object should handle anyOf 1`] = `"string | string[]"`;

exports[`create type from schema with complex object should handle nullable props 1`] = `"number | null"`;

exports[`create type from schema with complex object should handle oneOf 1`] = `"string | string[]"`;

exports[`create type from schema with refs should work in nested properties 1`] = `
"{
    a?: BaseItem;
}"
`;

exports[`create type from schema with refs should work referencing ref 1`] = `"BaseItem"`;
